src/diraccfg/cfg.py

```diff
@@ -934,10 +927,10 @@ def loadFromBuffer(self, data):
                 elif line[index] == "}":
                     try:
                         currentLevel = levelList.pop()
-                    except IndexError:
+                    except IndexError as e:
                         raise ValueError(
                             "The cfg file seems to close more sections than it opens (i.e. to many '}' vs '{'"
-                        )
+                        ) from e
```

Comment:
I thought the `from` was implicit ?

Reply:
The implicit is `from None` which implies there was an exception in the error handling. In this case I think we want `from [PREVIOUS EXCEPTION]` as we're transforming the original exception to something more meaningful.

See https://docs.python.org/3/tutorial/errors.html#exception-chaining

Reply:
Though actually maybe `from None` is more sensible...

Reply:
Yes, the message is the main observable difference but I think the message is actually very meaningful for expressing what is going on. I've put some examples with which use this preamble:

```python
In [1]: def commit():
   ...:     raise Exception("Failed to commit")
   ...:
   ...: def rollback():
   ...:     raise Exception("Failed to rollback, something is really bad!")
   ...:
   ...: config = {"user": "myuser"}
```

`from None` prevents exception chaining from happening as the original error is a meaningless implementation detail that shouldn't concern users of the code:

```python
In [2]: try:
   ...:     username = config["username"]
   ...: except KeyError:
   ...:     raise Exception("Config does not contain 'username'") from None
   ...:
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
Input In [2], in <cell line: 1>()
      2     username = config["username"]
      3 except KeyError:
----> 4     raise Exception("Config does not contain 'username'") from None

Exception: Config does not contain 'username'
```

`from [PREVIOUS EXCEPTION]` means that you've annotated the previous failure with more information:

```python
In [3]: try:
   ...:     commit()
   ...: except Exception as e:
   ...:     raise Exception("Failed to insert file into database") from e
   ...:
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
Input In [3], in <cell line: 1>()
      1 try:
----> 2     commit()
      3 except Exception as e:

Input In [1], in commit()
      1 def commit():
----> 2     raise Exception("Failed to commit")

Exception: Failed to commit

The above exception was the direct cause of the following exception:

Exception                                 Traceback (most recent call last)
Input In [3], in <cell line: 1>()
      2     commit()
      3 except Exception as e:
----> 4     raise Exception("Failed to insert file into database") from e

Exception: Failed to insert file into database
```

If you don't explicitly use `from` then it's implying the second exception happened in an unexpected place and there are either two issues (or the `except:` block is buggy):

```python
In [4]: try:
   ...:     commit()
   ...: except Exception as e:
   ...:     rollback()
   ...:     raise Exception("Failed to insert file into database") from e
   ...:
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
Input In [4], in <cell line: 1>()
      1 try:
----> 2     commit()
      3 except Exception as e:

Input In [1], in commit()
      1 def commit():
----> 2     raise Exception("Failed to commit")

Exception: Failed to commit

During handling of the above exception, another exception occurred:

Exception                                 Traceback (most recent call last)
Input In [4], in <cell line: 1>()
      2     commit()
      3 except Exception as e:
----> 4     rollback()
      5     raise Exception("Failed to insert file into database") from e

Input In [1], in rollback()
      4 def rollback():
----> 5     raise Exception("Failed to rollback, something is really bad!")

Exception: Failed to rollback, something is really bad!
```

Reply:
I see, so it really is a semantic difference :-)